backend/protondrive/protondrive.go

```diff
@@ -0,0 +1,1069 @@
+		size:         size,
+		originalSize: nil,
+		id:           "",
+		modTime:      modTime,
+		mimetype:     "",
+		link:         nil,
+	}
+	return obj, nil
+}
+
+// Mkdir makes the directory (container, bucket)
+//
+// Shouldn't return an error if it already exists
+func (f *Fs) Mkdir(ctx context.Context, dir string) error {
+	_, err := f.dirCache.FindDir(ctx, f.sanitizePath(dir), true)
+	return err
+}
+
+// Rmdir removes the directory (container, bucket) if empty
+//
+// Return an error if it doesn't exist or isn't empty
+func (f *Fs) Rmdir(ctx context.Context, dir string) error {
+	folderLinkID, err := f.dirCache.FindDir(ctx, f.sanitizePath(dir), false)
+	if err == fs.ErrorDirNotFound {
+		return fmt.Errorf("[Rmdir] cannot find LinkID for dir %s (%s)", dir, f.sanitizePath(dir))
+	} else if err != nil {
+		return err
+	}
+
+	if err = f.pacer.Call(func() (bool, error) {
+		err = f.protonDrive.MoveFolderToTrashByID(ctx, folderLinkID, true)
+		return shouldRetry(ctx, err)
+	}); err != nil {
+		return err
+	}
+
+	f.dirCache.FlushDir(f.sanitizePath(dir))
+	return nil
+}
+
+// Precision of the ModTimes in this Fs
+func (f *Fs) Precision() time.Duration {
+	return time.Second
+}
+
+// DirCacheFlush an optional interface to flush internal directory cache
+// DirCacheFlush resets the directory cache - used in testing
+// as an optional interface
+func (f *Fs) DirCacheFlush() {
+	f.dirCache.ResetRoot()
+	f.protonDrive.ClearCache()
+}
+
+// Returns the supported hash types of the filesystem
+func (f *Fs) Hashes() hash.Set {
+	return hash.Set(hash.SHA1)
+}
+
+// About gets quota information
+func (f *Fs) About(ctx context.Context) (*fs.Usage, error) {
+	var user *proton.User
+	var err error
+	if err = f.pacer.Call(func() (bool, error) {
+		user, err = f.protonDrive.About(ctx)
+		return shouldRetry(ctx, err)
+	}); err != nil {
+		return nil, err
+	}
+
+	total := int64(user.MaxSpace)
+	used := int64(user.UsedSpace)
+	free := total - used
+
+	usage := &fs.Usage{
+		Total: &total,
+		Used:  &used,
+		Free:  &free,
+	}
+
+	return usage, nil
+}
+
+// ------------------------------------------------------------
+
+// Fs returns the parent Fs
+func (o *Object) Fs() fs.Info {
+	return o.fs
+}
+
+// Return a string version
+func (o *Object) String() string {
+	if o == nil {
+		return "<nil>"
+	}
+	return o.remote
+}
+
+// Remote returns the remote path
+func (o *Object) Remote() string {
+	return o.remote
+}
+
+// Hash returns the hashes of an object
+func (o *Object) Hash(ctx context.Context, t hash.Type) (string, error) {
+	if t != hash.SHA1 {
+		return "", hash.ErrUnsupported
+	}
+
+	if o.digests != nil {
+		return *o.digests, nil
+	}
+
+	// sha1 not cached
+	log.Println("sha1 not cached")
+	// we fetch and try to obtain the sha1 of the link
+	fileSystemAttrs, err := o.fs.protonDrive.GetActiveRevisionAttrsByID(ctx, o.ID())
+	if err != nil {
+		return "", err
+	}
+
+	if fileSystemAttrs == nil || fileSystemAttrs.Digests == "" {
+		return "", ErrFileSHA1Missing
```

Comment:
@ncw one last quick question, if the sha1 is somehow not present in the file attribute from the API, should I probably throw `hash.ErrUnsupported` instead of my custom error so that operations like Copy() can handle it better?

Reply:
If a given file doesn't have an sha1 then return empty string with no error - that's what the other backends do.

I'd probably write a debug also if you don't think it will be common.

Reply:
Thanks @ncw, I will proceed with the change and run the integration tests for the one last time!

Reply:
> If a given file doesn't have an sha1 then return empty string with no error - that's what the other backends do.

Fix in #7163